compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -379,15 +381,19 @@ fn extend_type_not_partial_eq<'tcx>(
         adts_without_partialeq: FxHashSet<Span>,
         /// The user has written `impl PartialEq for Ty` which means it's non-structual,
         /// but we don't have a span to point at, so we'll just add them as a `note`.
-        manual: Vec<Ty<'tcx>>,
+        manual: FxHashSet<Ty<'tcx>>,
         /// The type has no `PartialEq` implementation, neither manual or derived, but
         /// we don't have a span to point at, so we'll just add them as a `note`.
-        without: Vec<Ty<'tcx>>,
+        without: FxHashSet<Ty<'tcx>>,
+        /// If any of the subtypes has escaping bounds (`for<'a>`), we won't emit a note.
+        has_escaping_bound_vars: bool,
     }
 
     impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for UsedParamsNeedInstantiationVisitor<'tcx> {
         fn visit_ty(&mut self, ty: Ty<'tcx>) -> Self::Result {
-            if let ty::Adt(def, _args) = ty.kind() {
+            if ty.has_escaping_bound_vars() {
+                self.has_escaping_bound_vars = true;
+            } else if let ty::Adt(def, _args) = ty.kind() {
```

Comment:
I also believe that this visitor should not recur into fn-ptrs instead of bailing on bound vars.

Whether an adt implements `PartialEq` does not matter if it is inside of a trait object which does not implement `PartialEq` regardless.

Reply:
why do you set `has_escaping_bound_vars = true` when encountering trait objects?

also, please make it an exhaustive match on `ty.kind()` and skip fn ptrs (they always impl eq) and bail on trait objects (they never do, except for `dyn PartialEq` :skull: :thinking:) and generators. this means `has_escaping_bound_vars` should be unnecessary

Reply:
I was using that as a signal of "don't provide the suggestion". Can rename the field name or add another field.